src/field/FieldStepper.hh

```diff
@@ -0,0 +1,73 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2020 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file FieldStepper.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include "base/Macros.hh"
+#include "base/Types.hh"
+#include "FieldEquation.hh"
+
+#include "field/base/OdeArray.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * The base class of Runga-Kutta family steppers
+ *
+ * \tparam T A derived class.
+ */
+template<typename T>
+class FieldStepper
+{
+    using ode_type = OdeArray;
+
+  public:
+    // Construct with the equation of motion
+    inline CELER_FUNCTION FieldStepper(FieldEquation& equation);
+
+    // Commom methods
+    CELER_FUNCTION real_type stepper(real_type       h,
+                                     const ode_type  y,
+                                     const ode_type& dydx,
+                                     ode_type&       yout);
+
+    CELER_FUNCTION real_type distance_chord(const ode_type y,
+                                            const ode_type yout);
+
+    CELER_FUNCTION void ode_rhs(ode_type y, ode_type& yout);
+
+    // Static interfaces (Mandatory methods)
+    CELER_FUNCTION void ode_stepper(real_type       h,
+                                    const ode_type  y,
+                                    const ode_type& dydx,
+                                    ode_type&       yout)
+    {
+        static_cast<T*>(this)->ode_stepper(h, y, dydx, yout);
+    }
```

Comment:
I don't think you need a recursive pattern here since `T` only needs to access the equation and not any helper functions or complex interfaces. It's very difficult to tell what data is shared, and what's local, because:
- there are missing `_` indicating member data
- lots of methods aren't marked `const` when they should be
- lots of data is being passed around by mutable reference.

Please address all instances of these last three bullets and then I'll continue reviewing.

Reply:
T will be any Runge-Kutta family stepper.  I am going to test other steppers, such as RK-Fehlberg, RK-Nystrom and etc. in the next extension, which we use the common ode_stepper method.  So, let's leave like this, but consolidate codes later (i.e, no static polymorphism) if other steppers are not as attractive as the classical-4th-order-RK.